### PR TITLE
set e.f in the CallExp's for _d_critical*

### PR DESCRIPTION
**Pull request by thewilsonator**
e.f is null otherwise and will speed up semantic validation and simplify logic rather than having to go through `(cast(FuncDeclaration)(cast(VarExp)e.e1).var)`.

### CODE REVIEW

**Comment by WalterBright** [comment]
I don't see how it will speed up `CallExp.semantic()` as that function starts by setting `f` to `null`.

**Comment by thewilsonator** [comment]
Well maybe not for `dmd` but I need it in LDC to do [1]
```
 void visit(CallExp *e) override {
    // SynchronizedStatement is lowered to
    //    Critsec __critsec105; // 105 == line number
    //    _d_criticalenter(& __critsec105); <--
    //    ...                                 |
    //    _d_criticalexit( & __critsec105);   |
    // So we intercept it with the CallExp ----

    if (e->f->ident == Id::criticalenter) {
      e->error("cannot use 'synchronized' in @compute code");
      stop = true;
      return;
    }

    if (e->f->ident == Id::criticalexit) {
      stop = true;
      return;
    }
    ...
}  
```

otherwise I have to do `(cast(FuncDeclaration*)(cast(VarExp*)e->e1)->var` if I want to access `e->f` and this also ensures that I never have to check if `e->f == nullptr` which will therefore crash if I haven't handled a case where `e->f` is null, which is a good thing.

[1]: https://github.com/ldc-developers/ldc/blob/master/gen/semantic-dcompute.cpp#L186

**Comment by thewilsonator** [comment]
The same also applies to the `_d_monitor_e[nter|xit]` but classes are banned in dcompute anyway so that is no problem.

**Comment by thewilsonator** [comment]
Also the frontend semantic analysis is bypassed anyway by directly setting the type.

**Comment by thewilsonator** [comment]
@WalterBright ping. I have already done this for LDC. It would be nice to reduce the upstream diff for something trivial like this.

**Comment by thewilsonator** [comment]
Actually I think it would make more sense to just always assign `f` in the constructor for `CallExp`.

**Comment by andralex** [comment]
status of this?

**Comment by thewilsonator** [comment]
Merged as is in LDC. Correcting the constructors is possible but is more likely to have a performance impact. Rebase and cover the case of `synchronized(o)` above is probably the better option given only this seems to be the only AST inconsistency that I have come across.

**Comment by andralex** [comment]
@thewilsonator thanks! do you plan to look at the conflict?

**Comment by thewilsonator** [comment]
Done. I don't know what was causing the conflict, it merged conflict free locally. *shrug*

**Comment by andralex** [comment]
cc @WalterBright @MartinNowak please decide on this while still fresh

**Review by JinShil — DISMISSED** [review]

**Comment by thewilsonator** [comment]
The one autoteser failure was a network failure on the BSD 32 bit. 
It'd be nice to merge this to reduce the LDC diff.

**Comment by JinShil** [comment]
Please rebase and force push.  It will force the auto-tester to run again.

**Comment by JinShil** [comment]
@thewilsonator I know you're going to hate me, but I couldn't go through with it.  It just seems to brittle.

> Actually I think it would make more sense to just always assign `f` in the constructor for `CallExp`.

What would it take to do that?  I think that would make me more comfortable putting my name next to the "merge" button.

**Comment by thewilsonator** [comment]
Changes to multiple constructors in `UnaExp` because `CallExp` just forwards its ctors. You'd then have to update anywhere a UnaExp should be a CallExp. 

TL;DR more code churn for no gain.

**Comment by JinShil** [comment]
It appears you only need to add a single constructor to `CallExp`, because `f` is a member of `CallExp`.  I just implemented it myself, and it was only an additional 6 lines.

**Comment by thewilsonator** [comment]
Cool! code please ;)

The point of this pull is to reduce the diff between LDC and DMD. I don't care how's it done.
If we have to update it once that's fine, its the languishing and causing merge conflicts that's the problem. Feel free to close this and open yours, or just edit this one.

**Comment by JinShil** [comment]
Also, you might need to add a corresponding `create` method and add it to `expression.h`.

**Comment by thewilsonator** [comment]
That was why I decided to go with the simple approach of setting the fields directly. Adding a constructor is fine but it isn't used much except by the `create` function which are static, which requires updating the classname that is used everywhere else in the codebase i think. As is currently this is constrained to a single file, adding and appropriately using a `create` changes a large number of files.

**Comment by JinShil** [comment]
I hear you, but simple, or fewer lines isn't necessarily best.  IMO, it would better to make the constructors rather than the silly casting of an expression in the immediate following line just to set a property.  

Also, I'm not the only one here, if others disagree with me, they are more than welcome to merge this PR.  But if it's going to be my name there next to the merge notice, I need to have more confidence in the implementation.

**Comment by ibuclaw** [comment]
Or couldn't there be a helper function called `genCfuncCall` - just noting that every location updated is doing the exact same thing -> possibile refactoring opportunity.

**Comment by thewilsonator** [comment]
Hmm? Why is that still failing?
`dmd/statementsem.d(3970): Error: unrecognized declaration`

`error(c.loc, "cannot put catch statement inside `%s`", Token.toChars(sc.os.tok));`
The offending line is nowhere near the code I touched and I balanced my braces.
(probably smth. very obvious but am jetlagged atm)

**Comment by thewilsonator** [comment]
@ibuclaw @JinShil Refactored. Didn't reduce the line count by as much as I'd hoped, but its much easier to follow.

**Comment by JinShil** [comment]
@thewilsonator, The refactoring is rather gratuitous.  I've submitted an alternative at #7916 for your consideration.

**Comment by thewilsonator** [comment]
Closing in favour of #7916
